File: C++/NucleusLib/pnSceneObject/plSceneObject.cpp
#include "plSceneObject.h"

plSceneObject::plSceneObject(PlasmaVer pv) {
    DrawIntf = new plKey();
    SimIntf = new plKey();
    CoordIntf = new plKey();
    AudioIntf = new plKey();
    SceneNode = new plKey();
}

plSceneObject::~plSceneObject() {
    DrawIntf->UnRef();
    SimIntf->UnRef();
    CoordIntf->UnRef();
    AudioIntf->UnRef();
    SceneNode->UnRef();
}

void plSceneObject::read(hsStream *S) {
    plSynchedObject::read(S);

    DrawIntf->readRef(S);
    SimIntf->readRef(S);
    CoordIntf->readRef(S);
    AudioIntf->readRef(S);

    Interfaces.setSize(S->readInt());
    for (int i=0; i<Interfaces.getSize(); i++)
        Interfaces[i].readRef(S);
    Modifiers.setSize(S->readInt());
    for (int i=0; i<Modifiers.getSize(); i++)
        Modifiers[i].readRef(S);

    SceneNode->readRef(S);
}

void plSceneObject::write(hsStream *S) {
    plSynchedObject::write(S);

    DrawIntf->writeRef(S);
    SimIntf->writeRef(S);
    CoordIntf->writeRef(S);
    AudioIntf->writeRef(S);

    S->writeInt(Interfaces.getSize());
    for (int i=0; i<Interfaces.getSize(); i++)
        Interfaces[i].writeRef(S);
    S->writeInt(Modifiers.getSize());
    for (int i=0; i<Modifiers.getSize(); i++)
        Modifiers[i].writeRef(S);
    SceneNode->writeRef(S);
}

File: Java/Plasma/NucleusLib/pnSceneObject/plSceneObject.java
package Plamsa.NucleusLib.pnSceneObject;

import java.io.IOException;
import Plasma.CoreLib.hsStream;
import Plasma.CoreLib.hsTArray;
import Plasma.NucleusLib.pnKeyedObject.*;
import Plasma.NucleusLib.pnNetCommon.plSynchedObject;

public class plSceneObject extends plSynchedObject {
    public plKey DrawIntf = new plKey();
    public plKey SimIntf = new plKey();
    public plKey CoordIntf = new plKey();
    public plKey AudioIntf = new plKey();
    public hsTArray<plKeyRef> Interfaces = new hsTArray<plKeyRef>();
    public hsTArray<plKeyRef> Modifiers = new hsTArray<plKeyRef>();
    public plKey SceneNode = new plKey();;

    public void read(hsStream S) throws IOException {
        super.read(S);

        DrawIntf.readRef(S);
        SimIntf.readRef(S);
        CoordIntf.readRef(S);
        AudioIntf.readRef(S);
        Interfaces.read(S);
        Modifiers.read(S);
        SceneNode.readRef(S);
    }

    public void write(hsStream S) throws IOException {
        super.write(S);

        DrawIntf.writeRef(S);
        SimIntf.writeRef(S);
        CoordIntf.writeRef(S);
        AudioIntf.writeRef(S);
        Interfaces.write(S);
        Modifiers.write(S);
        SceneNode.writeRef(S);
    }
}

